src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1412,44 +1421,97 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
-                        if (argRank != paramRank) return false;
-
-                        // If both have rank 0, it could also be because their type info is ambiguous.
-                        if (argRank == 0 && paramRank == 0) return false;
+                        throw new Exception("unkown ast type");
                     }
+                    if (!DoesParamRankMatchArg(fep,p, t,argIndex)) return false;
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
+        /// <summary>
+        /// Check that a parameter's rank and type are compatible with the given argument.
+        /// This function besides checking rank directly, also does checks using the FEP attributes.
+        /// </summary>
+        /// <param name="fep">the function end point wrapper</param>
+        /// <param name="paramType">type of param on the fep</param>
+        /// <param name="argType">actual argument type</param>
+        /// <param name="paramIndex">parameter index in list of params to fep.</param>
+        /// <returns></returns>
+        private bool DoesParamRankMatchArg(ProtoCore.CLRFunctionEndPoint fep,Type paramType, Type argType, int paramIndex)
+        {
+            if (!paramType.Equals(argType))
+            {
+                //if both types are value types, let's
+                //let coercion figure this out.
+                if (paramType.IsValueType && argType.IsValueType)
+                {
+                    return true;
+                }
+
+                var argRank = GetRank(argType);
+                var paramRank = GetParamRank(fep,paramType, paramIndex);
+
+                //if the paramRank is an arbitrary rank array
+                //replication should not occur for this parameter
+                //(except in case of rep guides, but that has been ruled out previously).
+                if (paramRank == DSASM.Constants.kArbitraryRank)
+                {
+                    return true;
+                }
+                if (argRank != paramRank) return false;
+
+                // If both have rank 0, it could also be because their type info is ambiguous.
+                // or it could be fine!
+                if (argRank == 0 && paramRank == 0)
+                {
+                    EmitILComment("param and arg are both rank 0, type info may be ambiguous.");
+                    if (argType == typeof(DSASM.CLRStackValue))
+                    {
+                        EmitILComment("argument is from a replicate call, and param is rank 0, fallback to replication.");
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+
+        private static int GetParamRank(ProtoCore.CLRFunctionEndPoint fep, Type type, int paramIndex)
+        {
+            if (fep.procedureNode.ArgumentTypes.ElementAt(paramIndex).rank == DSASM.Constants.kArbitraryRank)
```

Comment:
Is this equivalent to checking for the `ArbitraryDimensionArrayImport` attribute on the parameter? If so, do we need the `ParamAttributes` property defined in the `CLRFunctionEndPoint` class?

Reply:
yes - we can remove the `ParamAttributes` for now, we can always add it back if we decide to reduce our use of protocore type system.